File: radar-commons-android/src/main/java/org/radarcns/util/SynchronizedReference.java
package org.radarcns.util;

import java.io.IOException;
import java.lang.ref.WeakReference;

public abstract class SynchronizedReference<T> {
    private WeakReference<T> ref;

    public SynchronizedReference() {
        ref = null;
    }

    public synchronized T get() throws IOException {
        T localValue = null;
        if (ref != null) {
            localValue = ref.get();
        }
        if (localValue == null) {
            localValue = compute();
            ref = new WeakReference<>(localValue);
        }
        return localValue;
    }

    protected abstract T compute() throws IOException;
}

File: radar-commons-android/src/main/java/org/radarbase/util/SynchronizedReference.kt
package org.radarbase.util

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.io.IOException
import java.lang.ref.WeakReference

class SynchronizedReference<T>(private val supplier: suspend () -> T) {
    private val mutex = Mutex()
    private var ref: WeakReference<T>? = null

    @Throws(IOException::class)
    suspend fun get(): T = mutex.withLock {
        ref?.get()
            ?: supplier().also { ref = WeakReference(it) }
    }
}
